contracts/epochs/FlowEpoch.cdc

```diff
@@ -432,6 +498,199 @@ access(all) contract FlowEpoch {
             FlowEpoch.account.storage.load<Bool>(from: /storage/flowAutomaticRewardsEnabled)
             FlowEpoch.account.storage.save(enabled, to: /storage/flowAutomaticRewardsEnabled)
         }
+        
+        access(all) fun emitEpochRecoverEvent(epochCounter: UInt64,
+            startView: UInt64,
+            stakingEndView: UInt64,
+            endView: UInt64,
+            numViewsInStakingAuction: UInt64,
+            numViewsInDKGPhase: UInt64,
+            nodeIDs: [String],
+            clusterAssignments: [[String]],
+            randomSource: String,
+            targetDuration: UInt64,
+            targetEndTime: UInt64,
+            clusterQCVoteData: [FlowClusterQC.ClusterQCVoteData],
+            dkgPubKeys: [String],
+            ) {
+                
+            /// Construct the identity table for the recovery epoch
+            let nodes: [FlowIDTableStaking.NodeInfo] = []
+            for nodeID in nodeIDs {
+                nodes.append(FlowIDTableStaking.NodeInfo(nodeID: nodeID))
+            }
+
+            let dkgPhase1FinalView = startView + numViewsInStakingAuction + numViewsInDKGPhase - 1
+            let dkgPhase2FinalView = startView + numViewsInStakingAuction + (2 * numViewsInDKGPhase) - 1
+            let dkgPhase3FinalView = startView + numViewsInStakingAuction + (3 * numViewsInDKGPhase) - 1
+
+            /// emit EpochRecover event
+            emit FlowEpoch.EpochRecover(
+                counter: epochCounter,
+                nodeInfo: nodes,
+                firstView: startView,
+                finalView: endView,
+                clusterAssignments: clusterAssignments,
+                randomSource: randomSource,
+                DKGPhase1FinalView: dkgPhase1FinalView,
+                DKGPhase2FinalView: dkgPhase2FinalView,
+                DKGPhase3FinalView: dkgPhase3FinalView,
+                targetDuration: targetDuration,
+                targetEndTime: targetEndTime,
+                clusterQCVoteData: clusterQCVoteData,
+                dkgPubKeys: dkgPubKeys,
+            )
+        }
+
+        /// Performs sanity checks for the provided epoch configuration. It will ensure the following;
+        /// - There is a valid phase configuration.
+        /// - All nodes in the node ids list have a weight > 0.
+        /// If the configuration is a valid configuration the the staking auction,
+        /// qc voting and dkg will be ended depending on the current epoch phase.
+        access(all) fun recoverEpochPreChecks(startView: UInt64,
```

Comment:
I would consider renaming this function (or moving the latter half that stops voting, DKG, or the staking auction to a separate function). 

To me, "sanity checks" implies that we are validating inputs but not changing state. Currently the function is both validating inputs and modifying state.

Reply:
https://github.com/onflow/flow-core-contracts/pull/420/commits/802d8284a8709fe3a118256d09baa070661f7b8b